### PR TITLE
feat(sessions): implement external-plan execution-binding pipeline (PR 1)

### PR DESCRIPTION
**Pull request by Szczepanov**
## Summary

- Implements PR 1 of the external-plan execution-binding pipeline for [issue #434](https://github.com/Szczepanov/adaptive-training-recommender/issues/434) as planned in [`docs/plans/h4-external-plan-execution-binding-pipeline.md`](docs/plans/h4-external-plan-execution-binding-pipeline.md).
- Closes the execution gap for external-plan sessions: enables actionable v4 imported sessions to construct a valid `SessionReferenceBinding` (`kind: 'external_plan'`) and save an immutable, content-addressed `ExecutionPrescription` write-once, allowing the session to be launched and tracked via `SessionRunner.tsx`.
- Enforces critical domain safety and version gating in `Home.tsx`:
  - Enforces `activeExternal && isV4Plan(activeExternal.plan)` to ensure legacy v1–v3 plans remain on their un-bound path.
  - Strictly requires `externalVerdict.decision === 'proceed'` and `template.id !== 'rest_01'` to bind `primarySession`. Sessions with `skip`, `defer`, or `scale` verdicts do not receive a launch binding (preventing execution of unscaled blocks on scaled recommendations until block-scaling is implemented).
- Adds `prepareExternalPlanSessionLaunch` in `sessionAuthoringService.ts` with defensive validation (`validateSessionDefinition`), deterministic hashing (`definitionHash`, `prescriptionHash`), and write-once persistence via `executionPrescriptionService.savePrescription`.
- Defensively rejects imported `isEvent` target-event sessions at the launch-preparation boundary. External events are advisory/fixed-activity inputs and can carry external verdict metadata while the planner recommends a different executable session; they must never become an external-plan `primarySession` through an ambiguous caller path.
- Explicitly verifies that volatile `createdAt` provenance is excluded from the canonical execution prescription hash payload (ADR-0023 D-MSNAP).
- Adds comprehensive unit test coverage in `sessionAuthoringService.test.ts` for idempotency, `createdAt` hash omission, summary overrides, event rejection, and defensive validation.
- Reworks the H4 handoff document to distinguish implemented PR-1 invariants from PR-2+ occurrence/bundle/reassessment work, and records compatibility with the Firestore validation deduplication merged in #441.

## Validation

- [x] `uv run pytest` (backend changes)
- [x] `cd app && npm run check` (frontend changes)
- [x] `cd app && npm run test:rules` (Firestore rules changes)
- [x] `cd app && npm run simulate:scenarios` (recommendation-engine changes)
- [ ] `uv run ruff check .` and `uv run mypy src/garmin_sync` (backend changes — no backend source files modified)
- [ ] `cd app && npm run visual:refresh` (material UI changes — no visual UI layout modified)

Results:
- `uv run pytest`: 782 passed in 6.84s.
- `cd app && npm run check`: 3,693 passed across 395 suites; 0 TypeScript errors, 0 ESLint errors; knowledge & workout catalogs validated.
- `cd app && npm run test:rules`: 190 passed across 12 test files in the local Firestore emulator, verifying rules budget and expression limits for `SessionReferenceBinding` with `kind: 'external_plan'`.
- `cd app && npm run simulate:scenarios` & `simulate:diff`: 39 scenarios simulated; no unintended drift.
- `cd app && node scripts/check-policy-drift.mjs origin/main`: passed; no decision-affecting engine files modified.
- `npx vitest run src/services/sessionAuthoringService.test.ts`: 9 passed after the event-advisory regression test was added.
- GitHub CI Pipeline #2564: passed on head `69378af`; GitHub tested synthetic merge `b9caae7` against current `main` `c24a87b` (including #441).

## Risk and reviewer guidance

Low risk: scoped strictly to v4 external-plan sessions with defensive validation and idempotency.
1. Inspect the safety gate in `Home.tsx:476-483`: verify that `primarySession` requires `isV4Plan(activeExternal.plan)`, `externalVerdict.decision === 'proceed'`, and non-rest templates.
2. Inspect `prepareExternalPlanSessionLaunch` in `sessionAuthoringService.ts`: verify that target events are rejected, `displayMetadata` is preserved, `createdAt` is omitted from the hash via `canonicalExecutionPrescriptionJson`, and no redundant `session_occurrences` records are created.
3. Reviewer confirmation that `firestore.rules` already supports `kind: 'external_plan'` with optional `occurrenceId`, and full emulator suite passes cleanly. PR #441 has already landed on `main` to deduplicate `primarySession`/`additionalSessions` recommendation-audit validation.

## Domain invariants

- [x] Firestore writes remain user-scoped under `users/{APP_USER_ID}/...`; execution prescriptions are saved under `users/{userId}/execution_prescriptions/{hash}`.
- [x] Calendar-date logic uses `Europe/Warsaw`; no UTC `toISOString().split('T')[0]` regressions introduced.
- [x] No credentials, tokens, service-account files, or raw health payloads are included.
- [x] Recommendation decision logic was not changed; `POLICY_VERSION` drift was still evaluated with `check-policy-drift.mjs` and no decision-affecting engine files were modified.

## Screenshots or recordings

Not applicable — non-visual logic and data-binding wiring; existing `MorningDecisionCard` launch button affords starting the workout.

## Summary by CodeRabbit

- **New Features**
  - External Plan V4 sessions can be prepared and launched for valid definitions when recommendations receive an actionable “proceed” verdict.
  - Session preparation preserves plan bindings, content-addressed prescriptions, metadata, and display details while preventing duplicate records.
  - Optional summary overrides are retained for external-plan sessions.

- **Bug Fixes**
  - “Scale” verdicts no longer create a primary session.
  - Invalid or advisory target-event plans are rejected before persistence.
  - Launch failures no longer interrupt recommendation generation.

- **Documentation**
  - Clarified launch eligibility, validation, safety behavior, and verification coverage.

### CODE REVIEW

File: docs/plans/h4-external-plan-execution-binding-pipeline.md [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+## Root cause
+
+`SessionSourceRef`'s `kind: 'external_plan'` variant is declared in `app/src/sessions/models.ts`
+but is **never actually constructed anywhere in the codebase** — not for a v4 intraday
+bundle's second member (the original H4 ask), and not even for today's single primary
+external-plan session. External-plan sessions (`external-plan@1..4`) have never been
+wired into the newer multidomain execution pipeline:
+
+```
```
[the remark this answers is not in the document]
Comment (Szczepanov) [reply docs/plans/h4-external-plan-execution-binding-pipeline.md:116]: Added `text` language tag to the code fence to satisfy markdownlint MD040.

File: docs/plans/h4-external-plan-execution-binding-pipeline.md [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+       const unsignedPrescription: ExecutionPrescription = {
+           schemaVersion: 1,
+           prescriptionHash: '',
+           sessionSource,
+           definitionHash,
+           blocks: definition.blocks,
+           createdAt: new Date().toISOString(),
+       };
+       const prescriptionHash = await hashExecutionPrescription(unsignedPrescription);
+       await executionPrescriptionService.savePrescription(userId, { ...unsignedPrescription, prescriptionHash });
```
[the remark this answers is not in the document]
Comment (Szczepanov) [reply docs/plans/h4-external-plan-execution-binding-pipeline.md]: Updated `unsignedPrescription` construction in `prepareExternalPlanSessionLaunch` to include `displayMetadata` (title, summary, intent, dominantModality, and duration) matching the catalog prescription path.

File: docs/plans/h4-external-plan-execution-binding-pipeline.md [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+   safety-restricted modality) or `defer` (recovery mode), or if the recommended template was
+   swapped to canonical Rest (`rest_01`). A `primarySession` binding must **NEVER** be created
+   for a skipped or deferred session.
+
+   Gate the launch preparation strictly on:
+   ```typescript
+   const verdict = recommendationWithPrescription.externalVerdict?.decision;
+   const isActionable = verdict === 'proceed' || verdict === 'scale';
+   const notRest = recommendationWithPrescription.template.id !== 'rest_01';
+   if (isActionable && notRest && externalContext && 'definition' in externalContext.session) {
```
[the remark this answers is not in the document]
Comment (Szczepanov) [reply docs/plans/h4-external-plan-execution-binding-pipeline.md]: Enforced runtime schema discriminator check ctiveExternal && isV4Plan(activeExternal.plan) in Home.tsx and updated the plan document and checklist accordingly. This ensures v1–v3 plans continue their existing un-bound behavior and only v4 plans can construct a primarySession.

File: docs/plans/h4-external-plan-execution-binding-pipeline.md [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+       }
+   }
+   ```
+
+3. **Behavior on `scale` verdict:**
+   For catalog sessions, `resolveWorkoutPrescription` produces `adjustedBlocks`. External-plan
+   sessions currently lack a block-level auto-scaling transformer (only
+   `session.scaling.reducedSummary` and volume/intensity scalar bounds exist). In PR 1,
+   `session.definition` blocks remain the authored blocks, while `displayMetadata.summary`
+   captures `verdict.scaledSummary` (passed via `summaryOverride` above) so the athlete and
```
[the remark this answers is not in the document]
Comment (Szczepanov) [reply docs/plans/h4-external-plan-execution-binding-pipeline.md]: Addressed: excluded scale verdicts from primarySession creation in Home.tsx and updated the plan and PR description. Since block-level scaling for external plans does not yet exist, launching would execute full unscaled blocks; therefore, primarySession is strictly restricted to proceed verdicts. For scale verdicts, the recommendation continues to display the scaled prescription and reduction guidance via ExternalVerdictBanner without an unscaled launch affordance.

File: app/src/services/sessionAuthoringService.ts [hunk 5]
```diff
@@ -147,3 +148,74 @@ export async function prepareAuthoredOccurrenceLaunch(
...
+export async function prepareExternalPlanSessionLaunch(
+    userId: string,
+    externalPlan: {
+        planId: string;
+        revision: number;
+        contentHash: string;
+        session: ExternalPlanSessionV4;
+    },
+    summaryOverride?: string,
+    now = new Date().toISOString(),
```
[the remark this answers is not in the document]
Comment (Szczepanov) [reply app/src/services/sessionAuthoringService.ts:173]: hashExecutionPrescription delegates directly to canonicalExecutionPrescriptionJson (sessionDefinitionHash.ts:38), which uses pickDefined(prescription, ['schemaVersion', 'sessionSource', 'definitionHash', 'blocks', 'displayMetadata']) and therefore already omits createdAt from the canonical hash payload by design (ADR-0023 D-MSNAP). To make this explicit and prevent regression, we have added clarifying documentation in sessionAuthoringService.ts and added an explicit unit test in sessionAuthoringService.test.ts verifying that two invocations with different createdAt timestamps produce the exact same prescriptionHash.

File: app/src/services/sessionAuthoringService.test.ts [hunk 6]
```diff
@@ -88,3 +89,122 @@ describe('prepareAuthoredOccurrenceLaunch (M3.3)', () => {
...
+        const first = await prepareExternalPlanSessionLaunch('u1', externalPlan);
+        const second = await prepareExternalPlanSessionLaunch('u1', externalPlan);
+        expect(second.binding.prescriptionHash).toBe(first.binding.prescriptionHash);
+    });
+
+    it('omits volatile createdAt from the prescription hash payload', async () => {
+        const externalPlan = makeV4ExternalPlan();
+        const first = await prepareExternalPlanSessionLaunch('u1', externalPlan, undefined, '2026-01-01T00:00:00.000Z');
+        const second = await prepareExternalPlanSessionLaunch('u1', externalPlan, undefined, '2026-12-31T23:59:59.999Z');
+        expect(second.binding.prescriptionHash).toBe(first.binding.prescriptionHash);
```
[the remark this answers is not in the document]
Comment (Szczepanov) [reply app/src/services/sessionAuthoringService.test.ts:239]: Implemented: updated services.prescription.savePrescription in sessionAuthoringService.test.ts to use a stateful mock store matching ExecutionPrescriptionService's write-once semantics (if (store.has(hash)) return;). Added explicit assertions in the idempotency test verifying that the stored record is retained upon duplicate preparation without being overwritten or replaced.

**Comment by Szczepanov** [comment]
Deep review follow-up completed on head `69378af`.

What I re-checked:
- the full PR diff and launch path (`Home` → authoring service → `ExecutionPrescription` → resolver → `SessionRunner`);
- v4 import/validation and immutable plan-revision resolution;
- adjudication semantics for `proceed` / `scale` / `skip` / `defer`;
- the separate fixed-activity/advisory treatment of imported target events;
- Firestore recommendation/prescription validation, including the rule-budget deduplication merged to `main` in #441;
- all existing review threads and current mergeability.

Additional fixes pushed during this pass:
- `3e7dc40` — reject `isEvent` external target events at `prepareExternalPlanSessionLaunch`. This closes an ambiguous-call-path edge case where event-advisory metadata can coexist with a different executable recommendation; an advisory event can no longer become `primarySession` through the launch helper.
- `83d245e` — add the regression test and keep the stateful write-once persistence assertions. The focused authoring-service suite is now 9 tests.
- `69378af` — rewrite the H4 handoff doc around the implemented PR-1 invariants, remove stale `proceed | scale` guidance, document the event exclusion, and separate PR-2+ occurrence/bundle/D-REASSESS work.

Validation:
- GitHub CI Pipeline #2564 is fully green on `69378af`.
- GitHub’s synthetic merge commit is `b9caae7`, with parents current `main` `c24a87b` (including #441) and this PR head `69378af`; therefore the rules refactor and this execution-binding change have been tested together without needing a noisy branch-sync merge commit.
- All inline review threads are resolved; no merge conflict is present and GitHub reports the PR mergeable.

I also updated the PR description to include the event-advisory guard, the 9-test focused-suite count, the current #441 integration context, and the PR-2+ scope boundary.

No remaining correctness blocker found in this PR’s intended PR-1 scope. The next architectural step remains source-appropriate external-plan occurrence tracking before independently launchable bundle members and D-REASSESS.

**Comment by Szczepanov** [comment]
Addressed in commit \ed9f1143\:

- **Atomic Firestore transaction:** In \ExecutionPrescriptionService.savePrescription\, replaced the separate \getDoc\/\setDoc\ sequence with \
unTransaction\. If the content-addressed prescription document already exists, the transaction reads it, verifies the content hash, and returns early without calling \	ransaction.set\. This guarantees write-once semantics and preserves the original document's \createdAt\ timestamp under concurrent calls.
- **Dedicated test coverage:** Added \executionPrescriptionService.test.ts\ with 9 unit tests simulating Firestore transactions under contention and concurrent race conditions, asserting that the earliest write remains stored.
- **Authoring service concurrency test:** Added a concurrent test in \sessionAuthoringService.test.ts\ asserting that concurrent \prepareExternalPlanSessionLaunch\ calls sharing a prescription hash preserve the earliest write.
- **Validation:** Full validation suite passes (\
pm run check\ with 3,704 unit tests across 396 suites, \pytest\ with 782 passed, ruff and typecheck clean).

File: app/src/services/executionPrescriptionService.ts [hunk 7]
```diff
@@ -52,19 +52,21 @@ export class ExecutionPrescriptionService {
...
-                throw new Error(`Stored prescription ${prescription.prescriptionHash} does not match its content hash`);
+        await runTransaction(this.db, async transaction => {
+            const existing = await transaction.get(ref);
+            if (existing.exists()) {
+                const persisted = existing.data() as ExecutionPrescription;
+                const persistedHash = await hashExecutionPrescription(persisted);
+                if (persistedHash !== prescription.prescriptionHash) {
+                    throw new Error(`Stored prescription ${prescription.prescriptionHash} does not match its content hash`);
+                }
+                return;
```
[the remark this answers is not in the document]
Comment (Szczepanov) [reply app/src/services/executionPrescriptionService.ts:72]: Clarified to formalize write-once first-commit semantics:

- **Immutable domain contract:** In accordance with ADR-0023 (\execution_prescriptions/{prescriptionHash}\ is content-addressed and write-once), Firestore security rules strictly prohibit updates (\llow update, delete: if false;\ at \irestore.rules:1190\). Attempting to replace or mutate an existing record when a client arrives with an earlier timestamp would be rejected by Firestore with \permission-denied\ on the update rule.
- **Contract & documentation:** Added JSDoc on \ExecutionPrescriptionService.savePrescription\ explicitly defining the write-once, first-commit semantics (create-if-absent) where the first transaction to commit stores the record, and subsequent or concurrent calls verify content hash integrity and safely no-op without mutating the store.
- **Test coverage:** Updated test titles in \executionPrescriptionService.test.ts\ and \sessionAuthoringService.test.ts\ to specify first-commit semantics, and added an explicit test case (\enforces first-commit immutability: arrival with earlier timestamp does not overwrite committed record\) asserting that once committed, subsequent transactions with different timestamps never overwrite the record.
